Modules/socketmodule.c

```diff
@@ -1719,18 +1719,32 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
     case AF_UNIX:
     {
         Py_buffer path;
+        PyObject *fspath;
         int retval = 0;
 
-        /* PEP 383.  Not using PyUnicode_FSConverter since we need to
-           allow embedded nulls on Linux. */
-        if (PyUnicode_Check(args)) {
-            if ((args = PyUnicode_EncodeFSDefault(args)) == NULL)
-                return 0;
-        }
-        else
+        fspath = PyOS_FSPath(args);
+        if (fspath) {
+            args = fspath;
+            /* PEP 383.  Not using PyUnicode_FSConverter since we need to
+               allow embedded nulls on Linux. */
+            if (PyUnicode_Check(args)) {
+                PyObject *encoded_path = PyUnicode_EncodeFSDefault(args);
+                Py_DECREF(args);
+                args = encoded_path;
+                if (args == NULL)
+                    return retval;
+            }
+        } else {
+            PyErr_Clear();
```

Comment:
It is not good to clear arbitrary exception (including KeyboardInterrupt, MemoryError and RecursionError). `PyOS_FSPath()` raises TypeError if the argument does not support the path protocol, so it should only clear TypeError. All other exceptions should be reported to the user.

TypeError can also be raised if `__fspath__()` returns not str or bytes or raises TypeError itself, and it would be better to not silence it, but it is not easy to do, so we can ignore it for now.